pkg/feed/gcstracker.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2025 Google LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package feed
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"io"
+	"log"
+	"os"
+	"time"
+
+	gcs "cloud.google.com/go/storage"
+	"github.com/google/oss-rebuild/pkg/rebuild/rebuild"
+	"github.com/google/oss-rebuild/pkg/rebuild/schema"
+	"github.com/pkg/errors"
+)
+
+const TrackedPackagesFile = "tracked.json.gz"
+
+const UpdatePollingInterval = 15 * time.Minute
+
+type GCSTracker struct {
+	obj             *gcs.ObjectHandle
+	localCache      string
+	lastUpdate      time.Time
+	trackedPackages map[rebuild.Ecosystem]map[string]bool
+}
+
+// GCSTracker implements Tracker by reading a list of tracked packages
+// from a GCS object containing compressed JSON. The JSON should be a map from
+// ecosystem (string) to a list of package names (string[]).
+// GCSTracker keeps a local cache of the object to speed up repeated runs.
+func NewGCSTracker(obj *gcs.ObjectHandle, cachePath string) Tracker {
+	return &GCSTracker{obj: obj, localCache: cachePath}
+}
+
+func logFailure(f func() error) {
+	if err := f(); err != nil {
+		log.Println(err)
+	}
+}
+
+func (tracker *GCSTracker) fetchFromGCS(ctx context.Context) error {
```

Comment:
i think we should eagerly pre-populate prior to the first request. and we need mutual exclusion given inits can race.